### DISCUSSION TITLE
Set a clear `.astro` directory structure

### DISCUSSION THREAD

**Discussion by florian-lefebvre** [withastro/roadmap#887] [Proposal]
### Body

# Summary

Create a clear structure in `.astro`, for core and integrations

# Background & Motivation

Back in February, I started working on https://github.com/withastro/astro/pull/9952, innocently thinking that it would be easy. The goal was initially to do some tsconfig magic, then I scoped it down to allow integrations to inject types, then it was still too much so I closed it.

I uncovered that it requires a few things, including a rework of the `.astro` dir structure. According to @Princesseuh, it was initially planned for 4.0 but got postponed because of the Dev Toolbar.

Anyway, why do we need a clear structure? With more and more core and integrations features relying on codegen, the risk of collisions is increasing. The DX is not so great, and types require adding references in `src/env.d.ts`.

The goal is to unify all of this in a coherent structure.

# Goals

- Create a convention for the `.astro` structure
- Refactor how core refers to / interacts with the `.astro` directory

## Non goals

- Allow integrations to hook into this new structure

# Example

```
.astro/
	types.d.ts // only references to other .d.ts. We keep this name for backward compatibility,
				  can be changed to something like astro.d.ts in Astro 5
	astro/
		content.d.ts // formerly types.d.ts
	integrations/
		<integration_name>/
			<dts_name>.d.ts
```

**Comment by Princesseuh** [comment]
Love this 😄

How would you imagine this `types.d.ts` be included into the TypeScript project? Using `env.d.ts` like right now? I know Nuxt uses two `tsconfig.json` and that you tried that and there was a bunch of issues, but maybe we can figure it out.

For reference, a way to include `types.d.ts` is required because TypeScript does not include hidden folders unless there's a direct reference.

**Reply by florian-lefebvre** [reply]
Yeah for now we would use the current `types.d.ts` to avoid breaking changes but the next step is actually getting a generated tsconfig to work. Then we can do a codemod to remove the explicit reference. But that's quite far ahead already

**Comment by matthewp** [comment]
Will this make it easier to do type generation?

**Reply by Princesseuh** [reply]
> Will this make it easier to do type generation?

Yes, essentially one of the reason why it's hard to provide a cool API for Integrations and stuff to do type gen is that they have to not only, no real place to put their types, but also they have to do the hooking up in `env.d.ts` themselves.

With this, every integration gets its own happy place to do codegen (not just types) and Astro takes care of the hooking up.

**Reply by florian-lefebvre** [reply]
Exactly! Eventually, we'll track types filenames and generate relative references. In the end, the user will only need a reference to `.astro/types.d.ts` from `src/env.d.ts`

**Reply by Princesseuh** [reply]
And if we end up doing the tsconfig inside .astro thing, then the user won't need to do anything 🤩

**Reply by matthewp** [reply]
Is providing APIs for doing typegen in scope of this proposal?

**Reply by Princesseuh** [reply]
> Is providing APIs for doing typegen in scope of this proposal?

No, this would go in a further RFC specific to codegen. This proposal aims to only inform about the shape of the codegen folder.

Unless you mean for core, in which case yes. The main post goals and non goals talk about this.

**Reply by florian-lefebvre** [reply]
Yeah typegen apis for integrations will be in a following rfc. That's because I tried to tackled this at once already, thinking it would be straight forward, and it was an absolute mess involving many parts in astro

**Comment by florian-lefebvre** [comment]
Changes have been made
